transport/tchannel/outbound_test.go

```diff
@@ -47,6 +48,108 @@ var newOutbounds = []func(*tchannel.Channel, string) transport.Outbound{
 	},
 }
 
+func TestOutboundHeaders(t *testing.T) {
+	tests := []struct {
+		context context.Context
+		headers transport.Headers
+
+		wantHeaders []byte
+		wantError   string
+	}{
+		{
+			context: yarpc.WithBaggage(context.Background(), "FOO", "Bar"),
+			wantHeaders: []byte{
+				0x00, 0x01,
+				0x00, 0x0B, 'c', 'o', 'n', 't', 'e', 'x', 't', '-', 'f', 'o', 'o', // context-foo
+				0x00, 0x03, 'B', 'a', 'r',
+			},
+		},
+		{
+			headers: transport.Headers{"contextfoo": "bar"},
+			wantHeaders: []byte{
+				0x00, 0x01,
+				0x00, 0x0A, 'c', 'o', 'n', 't', 'e', 'x', 't', 'f', 'o', 'o',
+				0x00, 0x03, 'b', 'a', 'r',
+			},
+		},
+		{
+			headers: transport.Headers{"Foo": "bar"},
+			wantHeaders: []byte{
+				0x00, 0x01,
+				0x00, 0x03, 'f', 'o', 'o',
+				0x00, 0x03, 'b', 'a', 'r',
+			},
+		},
+		{
+			headers:   transport.Headers{"context-foo": "bar"},
+			wantError: `application headers cannot start with "Context-"`,
+		},
+	}
+
+	for _, tt := range tests {
+		server := testutils.NewServer(t, nil)
+		defer server.Close()
+		hostport := server.PeerInfo().HostPort
+
+		server.GetSubChannel("service").SetHandler(tchannel.HandlerFunc(
+			func(ctx context.Context, call *tchannel.InboundCall) {
+				var headers []byte
+
+				err := tchannel.NewArgReader(call.Arg2Reader()).Read(&headers)
+				if assert.NoError(t, err, "failed to read request headers") {
+					assert.Equal(t, tt.wantHeaders, headers, "headers did not match")
+				}
+
+				var body []byte
+				err = tchannel.NewArgReader(call.Arg3Reader()).Read(&body)
+				if assert.NoError(t, err, "failed to read request body") {
+					assert.Equal(t, []byte("world"), body)
+				}
+
+				err = tchannel.NewArgWriter(call.Response().Arg2Writer()).
+					Write([]byte{0x00, 0x00})
+				assert.NoError(t, err, "failed to write response headers")
+
+				err = tchannel.NewArgWriter(call.Response().Arg3Writer()).
+					Write([]byte("bye!"))
+				assert.NoError(t, err, "failed to write response")
+			}))
+
+		for _, getOutbound := range newOutbounds {
+			out := getOutbound(testutils.NewClient(t, &testutils.ChannelOpts{
+				ServiceName: "caller",
+			}), hostport)
```

Comment:
(no action for this diff) should `Outbound` also have a `Stop` / `Close` method? If we create a TChannel for an outbound within YARPC, we should expose a way to shut it down (since there'll be long lived connections in the background etc)